## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [5]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?
df[df['category'] == 'consumables']['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
df['first'].value_counts().head(5)

first
Christopher    1090
Michael        1047
William         982
John            959
Jennifer        902
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top_clients = df['client_id'].value_counts().head(5).index.tolist()
top_clients

[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the client with the most entries order order?
df[df['client_id'] == top_clients[0]]['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping'] = df['unit_weight'] * df['qty']
df['shipping'] = df['shipping'].apply(lambda x: x * 7 if x > 50 else x * 10)
df



first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  shipping  
0                1  115164.00   5512.50  
1                0     523.95    312.90  
2                6     527.28    458.64  
3                3    1056.18    356.70  
4                1    3902.00   6500.20  
...            ...        ...       ...  
54634            8    2743.29    519.75  
54635            0    9709.73   3849.30  
54636            6   31188.50  13832.00  
54637            0     165.76  14143.36  
54638            5    3103.31  14939.19  

[54639 rows x 20 columns]

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['sub_plus_ship'] = df['subtotal'] + df['shipping']
df['total_price'] = df['sub_plus_ship'] + (df['sub_plus_ship'] * 0.0925)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number   subtotal  shipping  sub_plus_ship  \
0             7.50   105            1  115164.00   5512.50      120676.50   
1             1.49    21            0     523.95    312.90         836.85   
2             1.68    39            6     527.28    458.64         985.92   
3             1.23    29            3    1056.18    356.70        1412.88   
4            46.43    20            1    3902.00   6500.20       10402.20   
...            ...   ...          ...        ...       ...            ...   
54634         2.25    33            8    2743.29    519.75        3263.04   
54635        11.70    47            0    9709.73   3849.30       13559.03   
54636         4.16   475            6   31188.50  13832.00       45020.50   
54637        18.04   112            0     165.76  14143.36       14309.12   
54638         2.07  1031            5    3103.31  14939.19       18042.50   

         total_price  
0      131839.076250  
1         914.258625  
2        1077.117600  
3        1543.571400  
4       11364.403500  
...              ...  
54634    3564.871200  
54635   14813.240275  
54636   49184.896250  
54637   15632.713600  
54638   19711.431250  

In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['cost'] = df['unit_cost'] * df['qty'] + df['shipping']
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  shipping  sub_plus_ship    total_price  \
0                1  115164.00   5512.50      120676.50  131839.076250   
1                0     523.95    312.90         836.85     914.258625   
2                6     527.28    458.64         985.92    1077.117600   
3                3    1056.18    356.70        1412.88    1543.571400   
4                1    3902.00   6500.20       10402.20   11364.403500   
...            ...        ...       ...            ...            ...   
54634            8    2743.29    519.75        3263.04    3564.871200   
54635            0    9709.73   3849.30       13559.03   14813.240275   
54636            6   31188.50  13832.00       45020.50   49184.896250   
54637            0     165.76  14143.36       14309.12   15632.713600   
54638            5    3103.31  14939.19       18042.50   19711.431250   

           cost  
0      85597.05  
1        629.79  
2        765.18  
3       1077.35  
4       8663.60  
...         ...  
54634   2222.55  
54635  12095.92  
54636  41054.25  
54637  14280.00  
54638  16599.10  

[54639 rows x 23 columns]

In [14]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['cost']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

        subtotal  shipping  sub_plus_ship    total_price      cost  \
0      115164.00   5512.50      120676.50  131839.076250  85597.05   
1         523.95    312.90         836.85     914.258625    629.79   
2         527.28    458.64         985.92    1077.117600    765.18   
3        1056.18    356.70        1412.88    1543.571400   1077.35   
4        3902.00   6500.20       10402.20   11364.403500   8663.60   
...          ...       ...            ...            ...       ...   
54634    2743.29    519.75        3263.04    3564.871200   2222.55   
54635    9709.73   3849.30       13559.03   14813.240275  12095.92   
54636   31188.50  13832.00       45020.50   49184.896250  41054.25   
54637     165.76  14143.36       14309.12   15632.713600  14280.00   
54638    3103.31  14939.19       18042.50   19711.431250  16599.10   

             profit  
0      46242.026250  
1        284.468625  
2        311.937600  
3        466.221400  
4       2700.803500  
...             ...  
54634   1342.321200  
54635   2717.320275  
54636   8130.646250  
54637   1352.713600  
54638   3112.331250  

[54639 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
print(df[df['order_id'] == 2742071]['total_price'].sum())
print(df[df['order_id'] == 2173913]['total_price'].sum())
print(df[df['order_id'] == 6128929]['total_price'].sum())

152811.897075
162388.71929999997
923441.24045


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in top_clients:
    print(client, df[df['client_id'] == client]['total_price'].sum())

33615 8377308.572500001
66037 10259514.801275
46820 9743794.320474999
38378 12906550.880874999
24741 82268891.97985001


In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
df_summary = pd.DataFrame(columns=['client_id', 'total_units', 'total_shipping', 'total_revenue', 'total_profit'])
df_summary['client_id'] = top_clients
df_summary['total_units'] = df_summary['client_id'].apply(lambda x: df[df['client_id'] == x]['qty'].sum())  
df_summary['total_shipping'] = df_summary['client_id'].apply(lambda x: df[df['client_id'] == x]['shipping'].sum())
df_summary['total_revenue'] = df_summary['client_id'].apply(lambda x: df[df['client_id'] == x]['total_price'].sum())    
df_summary['total_profit'] = df_summary['client_id'].apply(lambda x: df[df['client_id'] == x]['profit'].sum())  
df_summary

client_id  total_units  total_shipping  total_revenue  total_profit
0      33615        64313      1828984.89   8.377309e+06  2.201995e+06
1      66037        43018      1395151.85   1.025951e+07  3.255032e+06
2      46820        75768      1601448.84   9.743794e+06  2.736603e+06
3      38378        73667      3429455.40   1.290655e+07  3.271830e+06
4      24741       239862      5126448.37   8.226889e+07  3.657999e+07

In [18]:
# Format the data and rename the columns to names suitable for presentation.
if 'total_units' in df_summary.columns:
    df_summary.rename(columns={'total_units': 'Total Units Purchased'}, inplace=True)
if 'total_shipping' in df_summary.columns:
    df_summary.rename(columns={'total_shipping': 'Total Shipping Price'}, inplace=True)
if 'total_revenue' in df_summary.columns:
    df_summary.rename(columns={'total_revenue': 'Total Revenue'}, inplace=True)
if 'total_profit' in df_summary.columns:
    df_summary.rename(columns={'total_profit': 'Total Profit'}, inplace=True)
df_summary.rename(columns={'client_id': 'Client ID', 'total_units': 'Total Units Purchased', 'total_shipping': 'Total Shipping Price', 'total_revenue': 'Total Revenue', 'total_profit': 'Total Profit'}, inplace=True)
# Define the money columns. 
money_columns = ['Total Shipping Price', 'Total Revenue', 'Total Profit']
for column in money_columns:
        if column in df_summary.columns:
            df_summary[column] = pd.to_numeric(df_summary[column], errors='coerce')
# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    if pd.notnull(value):
        return f"${value / 1000000:.2f}M"


# Apply the currency_format_millions function to only the money columns. 
for column in money_columns:
    if column in df_summary.columns:
        df_summary[column] = df_summary[column].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
df_summary.rename(columns={'Total Shipping Price': 'Total Shipping Price (Millions)', 'Total Revenue': 'Total Revenue (Millions)', 'Total Profit': 'Total Profit (Millions)'}, inplace=True)
df_summary


Client ID  Total Units Purchased Total Shipping Price (Millions)  \
0      33615                  64313                          $1.83M   
1      66037                  43018                          $1.40M   
2      46820                  75768                          $1.60M   
3      38378                  73667                          $3.43M   
4      24741                 239862                          $5.13M   

  Total Revenue (Millions) Total Profit (Millions)  
0                   $8.38M                  $2.20M  
1                  $10.26M                  $3.26M  
2                   $9.74M                  $2.74M  
3                  $12.91M                  $3.27M  
4                  $82.27M                 $36.58M

In [20]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
df_summary_sorted = df_summary.sort_values(by='Total Profit (Millions)', ascending=False)
df_summary_sorted

Client ID  Total Units Purchased Total Shipping Price (Millions)  \
4      24741                 239862                          $5.13M   
3      38378                  73667                          $3.43M   
1      66037                  43018                          $1.40M   
2      46820                  75768                          $1.60M   
0      33615                  64313                          $1.83M   

  Total Revenue (Millions) Total Profit (Millions)  
4                  $82.27M                 $36.58M  
3                  $12.91M                  $3.27M  
1                  $10.26M                  $3.26M  
2                   $9.74M                  $2.74M  
0                   $8.38M                  $2.20M

In [ ]:
# I got stuck towards the middle of the code and found that I was not calculationg the sales tax correctly
# I found that I was not able to run the code unless i coerced the errors to False. 
# Overall it was a fun learning experience!
